### PR TITLE
Implement resource accounting for pods with the Mesos scheduler

### PR DESCRIPTION
**Pull request by sttts**
This patch
- set limits (0.25 cpu, 64 MB) on containers which are not limited in pod spec (these are also passed to the kubelet such that it uses them for the docker run limits)
- sums up the container resource limits for cpu and memory inside a pod,
- compares the sums to the offered resources
- puts the sums into the Mesos TaskInfo such that Mesos does the accounting
  for the pod.
- parses the static pod spec and adds up the resources
- sets the executor resources to 0.25 cpu, 64 MB plus the static pod resources
- sets the cgroups in the kubelet for system containers, resource containers and docker to the one
  of the executor that Mesos assigned
- adds scheduler parameters `--default-container-cpu-limit` and `--default-container-mem-limit`.

The containers themselves are resource limited by the Docker resource limits which
the kubelet applies when launching them.

Fixes mesosphere/kubernetes-mesos#68 and mesosphere/kubernetes-mesos#304

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
Hi @sttts, this patch is already a good step forward.  However, setting the parent cgroup is critical to both the accounting and resource isolation that the Mesos is designed to manage.

Regarding cgroups/cpu-shares, as-is the slave will allocate the computed amount of shares to a group for the _executor's_ process hierarchy, while the pod container also gets some amount of shares.  The aggregate effect is to skew the relative shares allocated to other (non-Kubernetes) executor containers.

With memory, the problem is more pernicious.  Here, unintended overcommitment could cause memory pressure to rise, forcing the kernel to start killing processes.

In short, we should pass the cgroup parent to the kubelet ASAP.

**Comment by sttts** [comment]
Does our executor have its own cgroup already? I would look into it, but need some assistance. Have no practical cgroup development experience.

**Comment by ConnorDoyle** [comment]
@sttts sorry for the delay.  Yes, the Mesos containerizer sets up executor cgroups already.  There might be some tricks required to discover the executor's parent cgroup.  This is somewhat a game of [fortunately, unfortunately](https://en.wikipedia.org/wiki/Fortunately,_Unfortunately) but it appears do-able.

**cgroup management in Mesos:**

Using `mesos-execute` to illustrate:

```
mesos-execute \
  --command="env | grep MESOS > /tmp/executor-env" \
  --master=127.0.0.1:5050 \
  --name=me
```

```
MESOS_DIRECTORY=/tmp/mesos/0/slaves/20150714-185421-16777343-5050-15267-S0/frameworks/20150714-185421-16777343-5050-15267-0003/executors/me/runs/bee3bebe-69e6-4dc5-9bf8-72cbb0af375a
MESOS_EXECUTOR_ID=test
MESOS_NATIVE_LIBRARY=/usr/local/lib/libmesos-0.22.1.so
MESOS_SLAVE_PID=slave(1)@127.0.0.1:5050
MESOS_FRAMEWORK_ID=20150714-185421-16777343-5050-15267-0003
MESOS_CHECKPOINT=0
MESOS_SLAVE_ID=20150714-185421-16777343-5050-15267-S0
```

The last name-part of the value for `MESOS_DIRECTORY` is actually the container ID, which is all we need to infer the parent cgroup.  Making some "safe" assumptions based on the current implementation (and also assuming the `--cgroups_hierarchy` and `--cgroups_root` slave flags are left at their respective defaults) we should find `/sys/fs/cgroup/mesos/{subsystem}/{container-id}` for each enabled subsystem (`cpu`, `memory`, etc.)

**Influencing cgroups for Docker containers:**

So, playing around with Docker's `--cgroup-parent` flag a bit (using Docker version 1.7):

`docker run --cgroup-parent test -d busybox sleep 1000`

yields cgroups `/sys/fs/cgroup/{subsystem}/test/`.

Curiously, adding an "absoute" parent did not do what I expected.

`docker run --cgroup-parent /my/cool/test -d busybox sleep 1000`

yields cgroups `/sys/fs/cgroup/{subsystem}/my/cool/test/`.

**Putting it all together:**

We should be able to read the container ID from the `$MESOS_DIRECTORY` environment variable when spinning up the executor and then set the cgroup parent for the kubelet executor to `"mesos/{container-id}"`.

**Summary:**

While this approach should work for most cases, we should figure out a "better" way to expose these values in subsequent versions of Mesos.  Many thanks to @tnachen for the [beta](https://en.wikipedia.org/wiki/Beta_(climbing)).

**Comment by sttts** [comment]
Will mesos increase and decrease the cpu/mem values of the respective cgroups automatically, when new mesos tasks are deployed with that executor or when they terminate? Or do we have to change resources of the executor itself?

**Comment by ConnorDoyle** [comment]
It's automatic. Mesos (via the cgroups/* isolators) will update() the resources for the top level of the hierarchy as tasks are launched or become terminal. In our case, every task is a Kubernetes pod.

**Comment by sttts** [comment]
Ok, sounds great. Will work on that tomorrow.

**Comment by ConnorDoyle** [comment]
One more note, we did find some oddities in cgroup naming using Docker 1.6, but that could have also been related to interaction with systemd on that host.  Needs further testing.

**Comment by sttts** [comment]
Any reason not to use `/proc/<executor_pid>/cgroup`?

**Comment by sttts** [comment]
To answer my own question: the cgroup name for each subsys might differ. So one would also need some heuristics to find the proper one provided by mesos.

**Comment by sttts** [comment]
The kubelet also does cgroup magic it seems (176 is the executor):
```bash
$ cat /proc/176/cgroup 
10:net_prio:/
9:perf_event:/kubelet
8:net_cls:/
7:freezer:/kubelet
6:devices:/kubelet
5:memory:/kubelet
4:blkio:/kubelet
3:cpuacct:/kubelet
2:cpu:/kubelet
1:cpuset:/kubelet
```

**Comment by sttts** [comment]
@ConnorDoyle ptal

File: contrib/mesos/pkg/executor/service/service.go [hunk 1]
```diff
@@ -70,6 +71,24 @@ type KubeletExecutorServer struct {
...
+		log.V(2).Infof("cannot derive executor's cgroup because MESOS_DIRECTORY is empty")
+		return ""
+	}
+
+	containerId := path.Base(mesosDir)
+	if containerId == "" {
+		log.V(2).Infof(`cannot derive executor's cgroup from MESOS_DIRECTORY="%v"`, mesosDir)
+		return ""
+	}
+	cgroupRoot := fmt.Sprintf("/mesos/%v", containerId)
```
Comment (ConnorDoyle) [inline contrib/mesos/pkg/executor/service/service.go]: Make the `"/mesos/` part configurable?

**Comment by sanjana-bhat** [comment]
Fixes https://github.com/mesosphere/kubernetes-mesos/issues/387

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 2]
```diff
@@ -33,8 +33,8 @@ import (
 )
 
 const (
-	DefaultContainerCpus = 0.25 // initial CPU allocated for executor
-	DefaultContainerMem  = 64   // initial MB of memory allocated for executor
+	DefaultContainerCpus = 0.25                   // initial CPU allocated for executor
+	DefaultContainerMem  = 64.0 * 1024.0 * 1024.0 // initial MB of memory allocated for executor
```
Comment (ConnorDoyle) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: ~64 million megabytes? Either the units in the comment or the value needs to be changed here :)

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 3]
```diff
@@ -209,39 +212,63 @@ func (t *T) Reset() {
...
 	)
 	for _, resource := range offer.Resources {
 		if resource.GetName() == "cpus" {
-			cpus = *resource.GetScalar().Value
+			offeredCpus = *resource.GetScalar().Value
 		}
 
 		if resource.GetName() == "mem" {
-			mem = *resource.GetScalar().Value
+			offeredMem = *resource.GetScalar().Value * 1024.0 * 1024.0
```
Comment (ConnorDoyle) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: The value from the Mesos resource offer should be interpreted as megabytes.

Comment (ConnorDoyle) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: n/m, looks like you're converting all of the values to bytes.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go (hunk 2 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: K8s uses bytes everywhere. To reuse the limit extraction code I do this as well everywhere and divide by 1024^2 when comparing to the offer values.

File: contrib/mesos/pkg/executor/service/service.go (hunk 1 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: Sure. Are there environments where the cgroup name is changed?

Comment (jdef) [inline contrib/mesos/pkg/executor/service/service.go]: +1 for configurability, perhaps also via env var? defaulting to /mesos is fine. would be nice to see "/mesos" as a constant as well.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go (hunk 2 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: k8s has a nice "resource" package that defines a Quantity type: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/resource/quantity.go

perhaps this could help to streamline things a bit?

File: contrib/mesos/pkg/executor/service/service.go [hunk 4]
```diff
@@ -70,6 +71,24 @@ type KubeletExecutorServer struct {
...
+	// derive our cgroup from MESOS_DIRECTORY environment
+	mesosDir := os.Getenv("MESOS_DIRECTORY")
+	if mesosDir == "" {
+		log.V(2).Infof("cannot derive executor's cgroup because MESOS_DIRECTORY is empty")
+		return ""
+	}
+
+	containerId := path.Base(mesosDir)
+	if containerId == "" {
+		log.V(2).Infof(`cannot derive executor's cgroup from MESOS_DIRECTORY="%v"`, mesosDir)
```
Comment (jdef) [inline contrib/mesos/pkg/executor/service/service.go]: fyi, `%q` works the same as `"%v"` here

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 5]
```diff
@@ -160,12 +160,15 @@ func (t *T) FillFromDetails(details *mesos.Offer) error {
 		panic("offer details are nil")
 	}
 
-	log.V(3).Infof("Recording offer(s) %v against pod %v", details.Id, t.Pod.Name)
+	cpu := t.computeResourceLimit("cpu", DefaultContainerCpus)
+	mem := t.computeResourceLimit("memory", DefaultContainerMem)
+
+	log.V(3).Infof("Recording offer(s) %v against pod %v: cpu: %f, mem: %f", details.Id, t.Pod.Name, cpu, mem)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: if we use k8s resource.Quantity it has nice output funcs for printing quants in a pretty human readable way.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 6]
```diff
@@ -209,39 +212,63 @@ func (t *T) Reset() {
...
+				value := limit.Value()
+				if value > 1000000 { // avoid overflow with MilliValue
+					x += float64(value)
+				} else {
+					x += float64(limit.MilliValue()) / 1000.0
+				}
+			}
+		}
+	}
+	if x == 0.0 {
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: actually if a pod has a container that doesn't specify limits, then the limits for the entire pod are unbounded.

see (and perhaps reuse?) https://github.com/GoogleCloudPlatform/kubernetes/blob/a8269e38c9e2bf81ba18cd6420e2309745d5b0b9/pkg/resourcequota/resource_quota_controller.go#L248

Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: Then the question is how to handle that. We cannot pass "unbounded" to Mesos I guess.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go (hunk 2 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: The resource type is only helpful when parsing and printing quantities. It does not support help with conversions.

File: contrib/mesos/pkg/executor/service/service.go (hunk 4 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: fixed.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go (hunk 6 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: Reusing the resourcequota functions now to extract limits. Otherwise, I have unified all the scheduler resource logic to use MB, i.e. the mesos memory unit. This simplifies the file a lot IMO.

**Comment by sttts** [comment]
@jdef ptal

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 7]
```diff
@@ -371,8 +375,8 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
 		log.Infof("Detected %d staticPods in Configuration.", numberStaticPods)
 
 		info.Resources = []*mesos.Resource{
-			mutil.NewScalarResource("cpus", float64(numberStaticPods)*podtask.DefaultContainerCpus),
-			mutil.NewScalarResource("mem", float64(numberStaticPods)*podtask.DefaultContainerMem),
+			mutil.NewScalarResource("cpus", float64(numberStaticPods)*float64(podtask.DefaultContainerCpus)),
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: we should do one of two things here:
(a) properly calculate the resource requirements of these static pods
(b) file an issue to accomplish (a); and leave a TODO(sttts) here

Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: um.. didn't see there was already a TODO

Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: can we make the following adjustment as part of this PR:

1. info.Resources should allocate some cpu and memory to account for the fact that the kubelet-executor and kube-proxy are running on the slave; should happen regardless of whether there are static pods. something like DefaultExecutorMem and DefaultExecutorCPU.
1. if there are static pods, the resources from the above should be included in the calculation

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 8]
```diff
@@ -26,15 +26,20 @@ import (
...
 	mutil "github.com/mesos/mesos-go/mesosutil"
 )
 
+type MegaBytes float64
+type CPUShares float64
+
 const (
-	DefaultContainerCpus = 0.25 // initial CPU allocated for executor
-	DefaultContainerMem  = 64   // initial MB of memory allocated for executor
+	DefaultContainerCpus = CPUShares(0.25) // initial CPU allocated for executor
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: since these are really tiny numbers for pods whose resources are supposed to be unbounded, do you think it's worth logging a warning? or info? to let the user know about the minuscule resources we're giving to them?

**Comment by jdef** [comment]
couple of minor nits. otherwise lgtm.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go (hunk 8 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: In fact using the default values is not enough. We have to patch the pods' limits to make the kubelet and Docker apply them. Otherwise, an umlimited pod might explode our resources limits in the executors mem cgroup.

**Comment by sttts** [comment]
Rebased to master.

@jdef ptal

File: contrib/mesos/pkg/archive/zip.go [hunk 9]
```diff
@@ -70,24 +70,24 @@ func ZipWalker(zw *zip.Writer) filepath.WalkFunc {
...
 		if !info.IsDir() {
-			numberManifests++
+			paths = append(paths, path)
 		}
 		return zipWalker(path, info, err)
 	}))
 
 	if err != nil {
-		return nil, 0, err
+		return nil, []string{}, err
```
Comment (jdef) [inline contrib/mesos/pkg/archive/zip.go]: why the []string{} instead of nil?

File: contrib/mesos/pkg/archive/zip.go [hunk 10]
```diff
@@ -70,24 +70,24 @@ func ZipWalker(zw *zip.Writer) filepath.WalkFunc {
...
 	} else if err = zw.Close(); err != nil {
-		return nil, 0, err
+		return nil, []string{}, err
```
Comment (jdef) [inline contrib/mesos/pkg/archive/zip.go]: ditto

Comment (sttts) [inline contrib/mesos/pkg/archive/zip.go]: right.

File: contrib/mesos/pkg/scheduler/podtask/pod_task.go [hunk 11]
```diff
@@ -213,35 +284,41 @@ func (t *T) AcceptOffer(offer *mesos.Offer) bool {
...
-	// therefore be compliant with expectations of mesos executors w/ respect to
-	// resource allocation and management.
-	//
-	// TODO(jdef): remove hardcoded values and make use of actual pod resource settings
-	if (cpus < DefaultContainerCpus) || (mem < DefaultContainerMem) {
-		log.V(3).Infof("not enough resources: cpus: %f mem: %f", cpus, mem)
+	// calculate cpu and mem sum over all containers of the pod
+	// TODO (@sttts): also support pod.spec.resources.limit.request
+	// TODO (@sttts): take into account the executor resources
+	podCopy := t.Pod // make copy to fill in the limits without modifying the pod
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: there's a DeepCopy API somewhere in k8s-land that actually does this correctly.
https://github.com/GoogleCloudPlatform/kubernetes/blob/e5c31da2ff29258240cd63f206922eeb18a6a27c/pkg/tools/etcd_helper.go#L484

in fact, since LimitPodCPU modifies the pod it's examining (and it's invoked earlier than here) we should probably be making copies sooner than this, yes?

Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/pod_task.go]: alternatively, don't make the change here; there's already an attempt to copy the pod object (or at least parts of it) in the scheduler (near where annotations are set, I think)

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 12]
```diff
@@ -351,38 +360,67 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
 
-		// Adjust the resource accounting for the executor.
-		// Currently each podTask accounts the default amount of resources.
-		// TODO(joerg84) adapt to actual resources specified by pods.
-		log.Infof("Detected %d staticPods in Configuration.", numberStaticPods)
+		// try to read pod files and sum resources
+		for _, podPath := range paths {
+			podJson, err := ioutil.ReadFile(podPath)
+			if err != nil {
+				return nil, nil, fmt.Errorf("error reading static pod spec: %v", err)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: attempt to continue instead?

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 13]
```diff
@@ -351,38 +360,67 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
+			}
+
+			pod := api.Pod{}
+			err = json.Unmarshal(podJson, &pod)
+			if err != nil {
+				return nil, nil, fmt.Errorf("error parsing static pod spec at %v: %v", podPath, err)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: ditto

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 14]
```diff
@@ -351,38 +360,67 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
+			}
 
-		info.Resources = []*mesos.Resource{
-			mutil.NewScalarResource("cpus", float64(numberStaticPods)*podtask.DefaultContainerCpus),
-			mutil.NewScalarResource("mem", float64(numberStaticPods)*podtask.DefaultContainerMem),
+			// TODO(sttts): allow unlimited static pods as well and patch in the default resource limits
+			unlimitedCPU := podtask.LimitPodCPU(&pod)
+			unlimitedMem := podtask.LimitPodMem(&pod)
+			if unlimitedCPU {
+				return nil, nil, fmt.Errorf("found static pod without limit on cpu resources: %v", podPath)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: ditto

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 15]
```diff
@@ -351,38 +360,67 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
+			}
+			if unlimitedMem {
+				return nil, nil, fmt.Errorf("found static pod without limit on memory resources: %v", podPath)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: ditto

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 12 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: not sure. Above we already exit when the path is invalid, then it should be fatal here too.

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 16]
```diff
@@ -70,6 +72,9 @@ const (
 	defaultReconcileInterval = 300    // 5m default task reconciliation interval
 	defaultReconcileCooldown = 15 * time.Second
 	defaultFrameworkName     = "Kubernetes"
+
+	executorCPUs = podtask.CPUShares(0.25) // initial CPU allocated for executor
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: wondering if these values should be "finely tuneable" via the scheduler config file

thoughts?

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 12 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: well, it's one thing if we can't access the entire collection of zip files; in that case we can't launch any static pods.

if someone fat-fingers a single pod desc, should that stop us from attempting to parse the rest? i'm not convinced. plus, someone may (for some reason) want to include a README in this dir. bombing out here precludes that (through, arguably, there's probably someplace for such a doc)

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 16 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: Yes, they should. Had planned that already.

**Comment by jdef** [comment]
overall looks pretty good. nicely done

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 12 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: In fact we need some more dynamic mechanism here: a watched directory. That's what the kubelet does. In this sense I totally agree, the error behavior should be different.

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 14 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: Here we get a problem: the pods are passed to the executor, consume memory and cpu. The executor only reserves 64MB for them. If they need more, docker does not stop them, but the executors cgroup will => boom

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 12 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: not super concerned that we don't have a directory watch implemented yet.
we can add it to the TODO list.

**Comment by sttts** [comment]
@jdef addressed all your comments. Add --default-container-cpu/mem-limit parameters and for that restructured all the resource code, especially moved it into their own scheduler/resource package.

ptal. Now everything is in I wanted to see in this PR.

**Comment by sttts** [comment]
Added documentation and squashed.

@jdef or @karlkfi Ready to be merged (it only changes the contrib/mesos directory).
